bitcoin/src/blockdata/script/num.rs

```diff
@@ -0,0 +1,570 @@
+
+use core::convert::TryInto;
+use core::fmt;
+// We specifically only implement ops that are implemented by `CScriptNum`. Furthermore we do not
+// implement ops from `CScriptNum` that use i64 as the `Rhs`, only those that use `Self` - this is
+// inline with the Rust standard integer types.
+use core::ops::{Add, AddAssign, BitAnd, BitAndAssign, Sub, SubAssign, Neg};
+
+#[cfg(all(test, mutate))]
+use mutagen::mutate;
+
+use crate::blockdata::locktime::absolute;
+use crate::blockdata::transaction::Sequence;
+
+/// A signed integer used in Bitcoin Script.
+///
+/// Bitcoin uses 32 bit integers however, for various reasons, the `CScriptNum` has a few quirks.
+/// Primarily, the `ScriptNum` is a signed 64 bit integer but the number of bytes in use is
+/// restricted in certain instances:
+///
+/// - Only 4 bytes (signed 32 bit integer) can be decoded from a script (excl. CSV and CLTV arguments).
+/// - For CSV and CLTV arguments up to 5 bytes can be decoded (unsigned 32 bit integer).
+/// - The whole 8 bytes can be encoded - this means one can encode a number that cannot be decoded.
+///
+/// From Bitcoin Core:
+///
+/// > Numeric opcodes (OP_1ADD, etc) are restricted to operating on 4-byte integers.
+/// > The semantics are subtle, though: operands must be in the range [-2^31 +1...2^31 -1],
+/// > but results may overflow (and are valid as long as they are not used in a subsequent
+/// > numeric operation). CScriptNum enforces those semantics by storing results as
+/// > an int64 and allowing out-of-range values to be returned as a vector of bytes but
+/// > throwing an exception if arithmetic is done or the result is interpreted as an integer.
+///
+/// ## Operands
+///
+/// - Operands must be in the range [-2^31 + 1 ... 2^31 - 1]
+/// - Note the range above _excludes_ i32::MIN. This because it requires 5 bytes to encode.
+/// - Operands (4 bytes) can be added together (up to 5 bytes), the resulting number _cannot_ be
+///   used as an operand but can appear in a script.
+/// - Note that the operands range _excludes_ i32::MIN
+///
+/// ## Minimal encoding
+///
+/// Encoding tries to use the least number of bytes, in doing so there can be ambiguity with
+/// negative numbers which must be explicitly resolved by pushing an extra byte.
+///
+/// The algorithm to encode `x` is as follows:
+///
+/// - First encode the absolute value of `x` as a little-endian array of bytes.
+/// - If the most significant byte is >= 0x80 and the value is positive, add a
+///   new zero-byte to make the significant byte < 0x80 again.
+/// - If the most significant byte is >= 0x80 and the value is negative, add a
+///   new 0x80 byte that will be popped off when decoding.
+/// - If the most significant byte is < 0x80 and the value is negative, add a
+///   now 0x80 byte, since it will be subtracted and interpreted as a negative
+///   when decoding.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ScriptNum(pub i64);
```

Comment:
Why is this `pub`? The entire point of the type was to avoid invalid values!

Reply:
Strictly speaking there are no "invalid values" for a ScriptNum, only values which the numeric opcodes will reject. (Amusingly, these opcodes will *output* such values.)

But I agree, we should hide the inner type and add a `new_from_i64` method or something with a big doccomment warning that if you provide out-of-range values you're probably going to crash your script, and that you probably don't want to use this constructor.